File: polymaker.pl
#!/usr/bin/perl
# main program starts here
########### You can change these settings if you need ###############
## Polygon style parameters (bias, factor, width) should be changed 
##           depending on your data.
# File names
my $data_fn = "ge_data.csv"; # input data file
my $kml_fn = "polygon.kml"; # output kml file

# Polygon style parameters
my $hgtBias = 0; # bias for data values
my $hgtFact = 5; # factor for polygon height
#  (polygon height[m] = ( value - bias ) * factor
my $polyWdt = 100; # width of polygon [m]
my $posShft = 0; # position shift [m]

# Polygon colors for min, max and transparency
my $rgb_min = "0000ff"; #hex of r,g,b for minimum value (default=0000ff: blue)
my $rgb_max = "ff0000"; #hex of r,g,b for maximum value (default=ff0000: red)
my $trp = "ff"; #transparency (default=ff: opaque)

#####################################################################

# Constant Pi
my $pi=3.14159265;

my $poly_fac = $polyWdt / 2 / 111000; #half width (meter to deg.)
my $poly_pos = $posShft / 111000; #position shift (meter to deg.)

# color hex
my $rr1 = substr($rgb_min, 0, 2);
my $gg1 = substr($rgb_min, 2, 2);
my $bb1 = substr($rgb_min, 4, 2);
my $rr2 = substr($rgb_max, 0, 2);
my $gg2 = substr($rgb_max, 2, 2);
my $bb2 = substr($rgb_max, 4, 2);

# hex to decimal
my $rd1 = hex($rr1);
my $gd1 = hex($gg1);
my $bd1 = hex($bb1);
my $rd2 = hex($rr2);
my $gd2 = hex($gg2);
my $bd2 = hex($bb2);

# Open input data file
open(IN,$data_fn) or die "$!";
my @lon; # longitude
my @lat; # latitude
my @val; # value
while(<IN>) {
    chomp($_); # delete LF
    my @buf = split(/,/, $_); # separate data by comma
    push @lon, $buf[0];
    push @lat, $buf[1];
    push @val, $buf[2];
}
close(IN);

# minimum and maximum value (for color setting)
my $val_min = min1( @val );
my $val_max = max1( @val );

# test print
#print "$val_min.\n";
#print "$val_max.\n";

# Initial GE viewpoint
# average, minimum, and maximum of lon and lat
my $lon_av = av1( @lon );
my $lat_av = av1( @lat );
my $lon_min = min1( @lon );
my $lon_max = max1( @lon );
my $lat_min = min1( @lat );
my $lat_max = max1( @lat );
# longitudinal and latitudinal ranges [m]
my $lon_range = 3 * ( $lon_max - $lon_min ) * 111000 * cos( abs($lat_av) * $pi / 180 );
my $lat_range = 3 * ( $lat_max - $lat_min ) * 111000;
my $v_range = 100;
# select larger value for $v_range
if ($lon_range > $lat_range) {
	$v_range = $lon_range;
}else{
	$v_range = $lat_range;
}

# create header kml text with default viewpoint
my $head_kml = 
"<?xml version=\"1.0\" encoding=\"UTF-8\"?>
<kml xmlns=\"http://earth.google.com/kml/2.1\">
<Document>
<name>Plot</name>
<open>1</open>
<LookAt>
<longitude>".$lon_av."</longitude>
<latitude>".$lat_av."</latitude>
<altitude>0</altitude>
<heading>0</heading>
<tilt>0</tilt>
<range>".$v_range."</range>
<gx:altitudeMode>relativeToSeaFloor</gx:altitudeMode>
</LookAt>
<Folder>
<name>Graph</name>
";

# kml text for data folder
my $data_kml = "
</Folder>
<Folder>
<name>Values</name>
";

# footer kml text
my $foot_kml = "
</Folder>
</Document>
</kml>
";

### creating polygon kml

my $poly_kml = ""; # polygon kml initialized
my $poly_each = ""; 

for (my $i = 0; $i <= $#val; $i++){

	$poly_each = ""; # each polygon kml initialized
	
	#Polygon color for each value
	if ($val_max == $val_min){  # if all values are the same, use color for max
	$rd = $rd2;
	$gd = $gd2;
	$bd = $bd2;
	}else{    # else, color numbers will be interpolated depending on their values
	$rd = int ( $rd1 + ( $val[$i] - $val_min ) / ( $val_max - $val_min ) * ( $rd2 - $rd1 ) );
	$gd = int ( $gd1 + ( $val[$i] - $val_min ) / ( $val_max - $val_min ) * ( $gd2 - $gd1 ) );
	$bd = int ( $bd1 + ( $val[$i] - $val_min ) / ( $val_max - $val_min ) * ( $bd2 - $bd1 ) );
	}

	#decimal to hex
	$rr = sprintf "%02x", $rd;
	$gg = sprintf "%02x", $gd;
	$bb = sprintf "%02x", $bd;
	

	#Polygon structure
	#4 positions (lon&lat [deg.]) are defined for polygon
	if (abs($lat[$i]) > 89) {  # eliminate north and south poles
		$lon1 = $lon[$i] - $poly_fac;
		$lon2 = $lon[$i] + $poly_fac;
	} else {    # longitudinal width should be corrected for the latitudinal effect
		$lon1 = $lon[$i] - $poly_fac / cos( abs($lat[$i]) * $pi / 180 );
		$lon2 = $lon[$i] + $poly_fac / cos( abs($lat[$i]) * $pi / 180 );
	}
	$lat1 = $lat[$i] - $poly_fac + $poly_pos;
	$lat2 = $lat[$i] + $poly_fac + $poly_pos;
	$val0 = ( $val[$i] - $hgtBias ) * $hgtFact; # polygon height calculation

	#output each polygon kml
	$poly_each ="
<Placemark>
<Style>
<LineStyle>
<color>00ffffff</color>
</LineStyle>
<PolyStyle>
<color>".$trp.$bb.$gg.$rr."</color>
<outline>1</outline>
</PolyStyle>
</Style>
<Polygon>
<extrude>1</extrude>
<altitudeMode>relativeToGround</altitudeMode>
<outerBoundaryIs>
<LinearRing>
<coordinates>\n"
.$lon1.",".$lat1.",".$val0."\n"
.$lon1.",".$lat2.",".$val0."\n"
.$lon2.",".$lat2.",".$val0."\n"
.$lon2.",".$lat1.",".$val0."\n"
.$lon1.",".$lat1.",".$val0."\n
</coordinates>
</LinearRing>
</outerBoundaryIs>
</Polygon>
</Placemark>
" ;

	# accumulate poly_each_kml to poly_kml
	$poly_kml = $poly_kml.$poly_each;
}  # end of for-loop


### creating value kml

my $value_kml = ""; # value kml initialized
my $value_each_kml = ""; 

for (my $i = 0; $i <= $#val; $i++){

	$value_each_kml = ""; # each value kml initialized

	# position of value display
	$lon1 = $lon[$i];
	$lat1 = $lat[$i] + $poly_pos;
	$val_num = $val[$i];
	$val0 = ( $val[$i] - $hgtBias ) * $hgtFact * 1.1; # slightly higher than the top of polygon
	
	$value_each_kml = "
<Placemark>
<name>".$val_num."</name>
<Style>
<IconStyle>
<Icon>
</Icon>
</IconStyle>
<LabelStyle>
<color>ff00ffff</color>
<scale>0.7</scale>
</LabelStyle>
</Style>
<Point>
<altitudeMode>relativeToGround</altitudeMode>
<coordinates>".$lon1.",".$lat1.",".$val0."</coordinates>
</Point>
</Placemark>
";

	# accumulate value_each_kml to value_kml
	$value_kml = $value_kml.$value_each_kml;
}  # end of for-loop


### Output all KML file
open(FH,"> $kml_fn");

print FH $head_kml;
print FH $poly_kml;
print FH $data_kml;
print FH $value_kml;
print FH $foot_kml;
close(FH);

######## end of main program  #########

# subroutine for max value
sub max1 {
    my $max = shift;    #first data
    foreach(@_){
        $max = $_ if( $max < $_ );    #replace with larger value
    }
    return( $max );
}

# subroutine for min value
sub min1 {
    my $min = shift;    #first data
    foreach(@_){
        $min = $_ if( $min > $_ );    #replace with smaller value
    }
    return( $min );
}

# subroutine for average value
sub av1 {
    my $av = 0;    #integration variable
    foreach(@_){
	$av = $av + $_;  #integrate
    }
    my $num = @_;        #number of elements
    $av = $av / $num;    #average
    return( $av );
}

File: polymaker.py
#!/usr/bin/python
# coding: UTF-8

# main program starts here

import math  # standard module

########### You can change these settings if you need ###############
## Polygon style parameters (bias, factor, width) should be changed 
##           depending on your data.
# File names
data_fn = "ge_data.csv" # input data file
kml_fn = "polygon.kml" # output kml file

# Polygon style parameters
hgtBias = 0.0 # bias for data values
hgtFact = 5.0 # factor for polygon height
#  (polygon height[m] = ( value - bias ) * factor
polyWdt = 100.0 # width of polygon [m]
posShft = 0.0 # position shift [m]

# Polygon colors for min, max and transparency
rgb_min = "0000ff" #hex of r,g,b for minimum value (default=0000ff: blue)
rgb_max = "ff0000" #hex of r,g,b for maximum value (default=ff0000: red)
trp = "ff" #transparency (default=ff: opaque)

#####################################################################

poly_fac = polyWdt / 2.0 / 111000.0 #half width (meter to deg.)
poly_pos = posShft / 111000.0 #position shift (meter to deg.)

# color hex
rr1 = rgb_min[0:2]
gg1 = rgb_min[2:4]
bb1 = rgb_min[4:6]
rr2 = rgb_max[0:2]
gg2 = rgb_max[2:4]
bb2 = rgb_max[4:6]

# hex to decimal
rd1 = int(rr1, 16)
gd1 = int(gg1, 16)
bd1 = int(bb1, 16)
rd2 = int(rr2, 16)
gd2 = int(gg2, 16)
bd2 = int(bb2, 16)

# Open input data file
lon = list(range(0))  # longitude
lat = list(range(0))  # latitude
val = list(range(0))  # value

f = open(data_fn, 'r')
while True:
  data = f.readline()
  data.rstrip('\n')  # remove LF
  if data == '':
    break
  l = [float(x.strip()) for x in data.split(',')]  # separate data by comma as numerical format
  lon = lon + [l[0]]
  lat = lat + [l[1]]
  val = val + [l[2]]
f.close()


# minimum and maximum value (for color setting)
val_min = min( val )
val_max = max( val )

# test print
#print(val_min)
#print(val_max)

# Initial GE viewpoint
# average, minimum, and maximum of lon and lat
lon_av = sum( lon ) / len( lon )
lat_av = sum( lat ) / len( lat )
lon_min = min( lon )
lon_max = max( lon )
lat_min = min( lat )
lat_max = max( lat )
# longitudinal and latitudinal ranges [m]
lon_range = 3.0 * ( lon_max - lon_min ) * 111000.0 * math.cos( abs(lat_av) * math.pi / 180.0 )
lat_range = 3.0 * ( lat_max - lat_min ) * 111000.0
v_range = 100.0
# select larger value for v_range
if lon_range > lat_range:
  v_range = lon_range
else:
  v_range = lat_range


# create header kml text with default viewpoint
head_kml = ('<?xml version="1.0" encoding="UTF-8"?>\n'
  '<kml xmlns="http://earth.google.com/kml/2.1">\n'
  '<Document>\n'
  '<name>Plot</name>\n'
  '<open>1</open>\n'
  '<LookAt>\n'
  '<longitude>' + str(lon_av) + '</longitude>\n'
  '<latitude>' + str(lat_av) + '</latitude>\n'
  '<altitude>0</altitude>\n'
  '<heading>0</heading>\n'
  '<tilt>0</tilt>\n'
  '<range>' + str(v_range) + '</range>\n'
  '<gx:altitudeMode>relativeToSeaFloor</gx:altitudeMode>\n'
  '</LookAt>\n'
  '<Folder>\n'
  '<name>Graph</name>\n')

# kml text for data folder
data_kml = ('</Folder>\n'
  '<Folder>\n'
  '<name>Values</name>\n')

# footer kml text
foot_kml = ('</Folder>\n'
  '</Document>\n'
  '</kml>\n')

### creating polygon kml

poly_kml = '' # polygon kml initialized
poly_each = '' 

for loni, lati, vali in zip(lon, lat, val):

  poly_each = '' # each polygon kml initialized
	
  #Polygon color for each value
  if val_max == val_min :  # if all values are the same, use color for max
    rd = rd2
    gd = gd2
    bd = bd2
  else :    # else, color numbers will be interpolated depending on their values
    rd = int ( rd1 + ( vali - val_min ) / ( val_max - val_min ) * ( rd2 - rd1 ) )
    gd = int ( gd1 + ( vali - val_min ) / ( val_max - val_min ) * ( gd2 - gd1 ) )
    bd = int ( bd1 + ( vali - val_min ) / ( val_max - val_min ) * ( bd2 - bd1 ) )

  #decimal to hex
  rr = format(rd, '02x')
  gg = format(gd, '02x')
  bb = format(bd, '02x')

  #Polygon structure
  #4 positions (lon&lat [deg.]) are defined for polygon
  if abs(lati) > 89.0 :  # eliminate north and south poles
    lon1 = loni - poly_fac
    lon2 = loni + poly_fac
  else :    # longitudinal width should be corrected for the latitudinal effect
    lon1 = loni - poly_fac / math.cos( abs(lati) * math.pi / 180.0 )
    lon2 = loni + poly_fac / math.cos( abs(lati) * math.pi / 180.0 )

  lat1 = lati - poly_fac + poly_pos
  lat2 = lati + poly_fac + poly_pos
  val0 = ( vali - hgtBias ) * hgtFact # polygon height calculation

  #output each polygon kml
  poly_each =('<Placemark>\n'
    '<Style>\n'
    '<LineStyle>\n'
    '<color>00ffffff</color>\n'
    '</LineStyle>\n'
    '<PolyStyle>\n'
    '<color>' + trp + bb + gg + rr + '</color>\n'
    '<outline>1</outline>\n'
    '</PolyStyle>\n'
    '</Style>\n'
    '<Polygon>\n'
    '<extrude>1</extrude>\n'
    '<altitudeMode>relativeToGround</altitudeMode>\n'
    '<outerBoundaryIs>\n'
    '<LinearRing>\n'
    '<coordinates>\n'
    + str(lon1) + ',' + str(lat1) + ',' + str(val0) + '\n'
    + str(lon1) + ',' + str(lat2) + ',' + str(val0) + '\n'
    + str(lon2) + ',' + str(lat2) + ',' + str(val0) + '\n'
    + str(lon2) + ',' + str(lat1) + ',' + str(val0) + '\n'
    + str(lon1) + ',' + str(lat1) + ',' + str(val0) + '\n'
    '</coordinates>\n'
    '</LinearRing>\n'
    '</outerBoundaryIs>\n'
    '</Polygon>\n'
    '</Placemark>\n')

  # accumulate poly_each_kml to poly_kml
  poly_kml = poly_kml + poly_each
# end of for-loop


### creating value kml

value_kml = '' # value kml initialized
value_each_kml = '' 

for loni, lati, vali in zip(lon, lat, val):

  value_each_kml = '' # each value kml initialized

  # position of value display
  lon1 = loni
  lat1 = lati + poly_pos
  val_num = vali
  val0 = ( vali - hgtBias ) * hgtFact * 1.1 # slightly higher than the top of polygon

  value_each_kml = ('<Placemark>\n'
    '<name>' + str(val_num) + '</name>\n'
    '<Style>\n'
    '<IconStyle>\n'
    '<Icon>\n'
    '</Icon>\n'
    '</IconStyle>\n'
    '<LabelStyle>\n'
    '<color>ff00ffff</color>\n'
    '<scale>0.7</scale>\n'
    '</LabelStyle>\n'
    '</Style>\n'
    '<Point>\n'
    '<altitudeMode>relativeToGround</altitudeMode>\n'
    '<coordinates>' + str(lon1) +',' + str(lat1) + ',' + str(val0) + '</coordinates>\n'
    '</Point>\n'
    '</Placemark>\n')

  # accumulate value_each_kml to value_kml
  value_kml = value_kml + value_each_kml
# end of for-loop


### Output all KML file
# all kml text
kml_all = head_kml + poly_kml + data_kml + value_kml + foot_kml

with open(kml_fn, mode='w') as f:
  f.write(kml_all)

######## end of main program  #########

